denlillemand/week2/index.html

```diff
@@ -0,0 +1,302 @@
+		    this.state = {
+			name: "",
+		    };
+		    this.onChange = this.onChange.bind(this);
+		    this.onSubmit = this.onSubmit.bind(this);
+		}
+
+		onChange(event) {
+		    this.setState({
+			name: event.target.value
+		    });
+		}
+
+		onSubmit(event) {
+		    this.props.onSubmitNamedList(this.state.name);
+		    this.setState({
+			name: ""
+		    });
+		}
+
+		render() {
+		    return(
+		    	<div>
+			    <input value={this.state.name} onChange={this.onChange} />
+			    <button onClick={this.onSubmit}>Submit list</button>
+			</div>
+		    );
+		}
+	    } 
+
+
+	    class Message extends React.Component {
+		constructor(props, context) {
+		    super(props, context);
+		    this.state = {};
+		    this.deleteMessage = this.deleteMessage.bind(this);
+		    this.archive = this.archive.bind(this);
+		}
+
+		deleteMessage(event) {
+		    this.props.deleteMessage(this.props.message.id);
+		}
+
+		archive(event) {
+		    let message = this.props.message;
+		    this.props.archive(message.id, message.archived);
+		}
+
+		render() {
+		    return(
+		    	<div>
+			    	{this.props.message.id}:
+				{this.props.message.text}:
+				{this.props.message.archived ? "ARCHIVED" : "NOT ARCHIVED"}
+				<button onClick={this.archive}>{this.props.message.archived ? "Unarchive" : "Archive"}</button>
+				<button onClick={this.deleteMessage}>Delete</button>
+			</div>
+		    
+		    );
+		}
+	    }
+	    Message.propTypes = {
+		deleteMessage: React.PropTypes.func,
+		message: React.PropTypes.object,
+		archive: React.PropTypes.func
+	    }
+
+	    class MainList extends React.Component {
+		constructor(props, context) {
+		    super(props, context);
+		    this.state = {};
+		    this.swapMessage = this.swapMessage.bind(this);
+		    this.archiveMessage = this.archiveMessage.bind(this);
+		    this.onSubmitNamedList = this.onSubmitNamedList.bind(this);
+		    this.deleteMessage = this.deleteMessage.bind(this);
+		    this.onSubmitMessage = this.onSubmitMessage.bind(this);
+		}
+
+		archiveMessage(id, currentList, isArchived) {
+		    var mutatedArray = this.state[currentList].map(function(message) {
+			if(message.id === id) {
+			    var clonedMessage = Object.assign({},message);
+			    clonedMessage.archived = !isArchived;
+			    return clonedMessage;
+			}
+			return message;
+		    });
+		    this.setState({
+			[currentList]: mutatedArray
+		    });
+		}
+
+		swapMessage(messageId, destinationList, currentList) {
+		    let messageIndex = -1;
+		    this.state[currentList].forEach(function(message, index, array) {
+			if(message.id === messageId) {
+			    messageIndex = index;
+			}
+		    });
+		    if(messageIndex > -1) {
+			let copiedCurrentList = this.state[currentList].slice();
+			let removedMessage = copiedCurrentList.splice(messageIndex, 1);
+			if(removedMessage[0].archived) {
+			    alert("Cannot remove an archived message");
+			} else {
+			    let copiedDestinationList = this.state[destinationList].slice();
```

Comment:
This will return a new array instance; a couple lines below you have `copiedDestinationList.concat(removedMessage)` which also returns a new array instance. Either use `slice` and then `push` to the newly created array; or `concat`  to the state array which returns a new array instance, hence preventing you from mutating state directly.